### PR TITLE
CMOS-301: Allow setting a path prefix

### PR DESCRIPTION
**Pull request by markspolakovs**
Also renamed the docs subpath to `/docs`, partly to avoid confusing routes and partly to avoid ridiculous URLs like `/cmos/cmos/cmos/0.1/index.html`.

### CODE REVIEW

**Comment by markspolakovs** [comment]
I've done a fair amount of testing (though I'll see about adding some integration tests), but it'd be good if you could give it a whack too.

**Comment by markspolakovs** [comment]
CI doesn't run integration tests so triggered a nightly run at https://github.com/couchbaselabs/observability/actions/runs/1806121948.

**Comment by markspolakovs** [comment]
Nightly run in https://github.com/couchbaselabs/observability/actions/runs/1807050337

**Comment by markspolakovs** [comment]
Nightly tests are struggling (see https://issues.couchbase.com/browse/CMOS-311), but the tests relating to this change specifically are passing at least.

File: docs/modules/ROOT/pages/configure-cmos.adoc [hunk 1]
```diff
@@ -16,6 +16,20 @@ As shown in the above diagram, you can customize the installation by overriding
...
 If you choose to disable a service, this is achieved by setting the corresponding environment variables of the docker container. The naming convention to be followed to disable a service is `DISABLE_X`, where `X` is the name of the service (e.g.,`DISABLE_ALERT_MANAGER`, `DISABLE_GRAFANA`, `DISABLE_JAEGER`, `DISABLE_LOKI`, `DISABLE_PROMETHEUS` and `DISABLE_WEBSERVER`).
 
+== Running behind a reverse proxy
+
+CMOS can be run behind a reverse proxy in your deployments.
+There are two ways of running it: on a subdomain (e.g. `cmos.example.com`), or on a sub-path (e.g. `example.com/cmos`).
+
+If running on a subdomain, you do not need to change any CMOS configuration, just point your reverse proxy at port 8080 on the CMOS container.
+
+If you are running on a sub-path, you will need to set the `CMOS_HTTP_PATH_PREFIX` environment variable to ensure all the services reflect this, for example `docker run -e "CMOS_HTTP_PATH_PREFIX=/cmos/" ...`.
```
Comment (redm2) [inline docs/modules/ROOT/pages/configure-cmos.adoc:26]: Remove trailing slash from example? (`=/cmos/`)

File: microlith/entrypoints/prometheus.sh [hunk 2]
```diff
@@ -19,6 +19,7 @@ export CB_MULTI_ADMIN_USER=${CB_MULTI_ADMIN_USER:-admin}
 export CB_MULTI_ADMIN_PASSWORD=${CB_MULTI_ADMIN_PASSWORD:-password}
 export CB_SERVER_AUTH_USER=${CB_SERVER_AUTH_USER:-Administrator}
 export CB_SERVER_AUTH_PASSWORD=${CB_SERVER_AUTH_PASSWORD:-password}
+export CMOS_HTTP_PATH_PREFIX=${CMOS_HTTP_PATH_PREFIX:-}
```
Comment (redm2) [inline microlith/entrypoints/prometheus.sh:22]: Worth explicitly setting the empty string "" here as default?

File: microlith/entrypoints/webserver.sh [hunk 3]
```diff
@@ -14,4 +14,10 @@
 # limitations under the License.
 
 set -ex
```
Comment (redm2) [inline microlith/entrypoints/webserver.sh:16]: Still want `-x` set?

File: microlith/nginx/nginx.conf.tmpl [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            }
+
+            # Static assets are served here and there's no easy way to change it
+            location {{ $subPath }}/ui/assets/ {
+                proxy_pass  http://localhost:7196;
+            }
+
+            # UI assumes the cbmultimanager REST API is on /api
+            location {{ $subPath }}/api/ {
+                # cbmm doesn't yet support path prefixes so strip it
```
Comment (redm2) [inline microlith/nginx/nginx.conf.tmpl:90]: There a CMOS to track this? (Might be worth adding it in, or removing 'yet')

File: microlith/run.sh [hunk 5]
```diff
@@ -33,18 +33,25 @@ function log() {
 }
 
 # Expose all nested config variables to make it simple to see
+if [[ "${CMOS_HTTP_PATH_PREFIX:-}" == */ ]]; then
```
Comment (redm2) [inline microlith/run.sh:36]: Would we prefer to just sanitize the input and remove an existing trailing slash ourselves? Simplifies things in some ways (for the user).

**Review by redm2 — APPROVED** [review]

File: docs/modules/ROOT/pages/configure-cmos.adoc (hunk 1 above)
Comment (markspolakovs) [reply docs/modules/ROOT/pages/configure-cmos.adoc:26]: Good catch!

File: microlith/entrypoints/webserver.sh (hunk 3 above)
Comment (markspolakovs) [reply microlith/entrypoints/webserver.sh:16]: webserver is simple enough that I think it's useful

**Pull request merged**